### PR TITLE
[RISC-V] Update TODOs for RiscV64 Machine and Architecture in Crossgen2

### PR DESCRIPTION
**Pull request by ashaurtaev**
Part of https://github.com/dotnet/runtime/issues/96438
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @yurai007 @sirntar @tomeksowi @Bajtazar @brucehoult

Merge after https://github.com/dotnet/runtime/pull/95980

### CODE REVIEW

**Review by clamp03 — APPROVED** [review]

**Review by gbalykov — APPROVED** [review]

**Review by t-mustafin — APPROVED** [review]

**Review by tomeksowi — APPROVED** [review]

**Comment by clamp03** [comment]
If you don't mind, please set to draft before #95980 is merged.

**Review by sirntar — APPROVED** [review]

**Review by wscho77 — APPROVED** [review]

**Comment by tomeksowi** [comment]
#95980 is merged, I think you can rebase and re-open for review.

**Comment by ashaurtaev** [comment]
@am11 After rebasing to the current main and installing the new dotnet-sdk-9.0.100-alpha.1.23615.4, I got the following errors:
```
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/GCInfoTypes.cs(157,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/GCInfoTypes.cs(173,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/GCInfoTypes.cs(189,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/GCInfoTypes.cs(207,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/GCInfoTypes.cs(223,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/TransitionBlock.cs(37,30): error CS0117: 'Machine' does not contain a definition for 'RiscV64' [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs(636,30): error CS1061: 'Machine' does not contain a definition for 'RiscV64' and no accessible extension method 'RiscV64' accepting a first argument of type 'Machine' could be found (are you missing a using directive or an assembly reference?) [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs(1420,38): error CS1061: 'Machine' does not contain a definition for 'RiscV64' and no accessible extension method 'RiscV64' accepting a first argument of type 'Machine' could be found (are you missing a using directive or an assembly reference?) [/home/runtime/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ILCompiler.Reflection.ReadyToRun.csproj]
    0 Warning(s)
    8 Error(s)
```
What would cause it and what can be done to fix it?

**Comment by ViktorHofer** [comment]
> Not sure if that coordination is still needed, since [sdk is using 8.0](https://github.com/dotnet/sdk/blob/eb47a3cf296746d8624124b5c002db9dc81bbc44/eng/Versions.props#L36) while runtime is at 7.0, and code flow is working fine. @ViktorHofer, thoughts on updating S.R.M to latest alpha?

cc @jaredpar @rainersigwald

**Comment by am11** [comment]
Ops sorry, the more relevant version is hardcoded to 6.0 here: https://github.com/dotnet/runtime/blob/86f25ad9bd314d5ffaca2c4aa18c56347497aaa7/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L22-L24

@ashaurtaev, could you push the change with `9.0.0-alpha.1.24064.3` only in `eng/Versions.props` for `SystemReflectionMetadataVersion` (since that one is internal to runtime). Lets see how it well bodes with the CI. :)

**Comment by ViktorHofer** [comment]
Which component needs a 9.0 version of SRM? If it's internal then that component might be better off using the live build of SRM.

Regardless, the HostModel version SRM version and others that are hardcoded should probably be updated to 8.0.0.

**Comment by ashaurtaev** [comment]
@am11 I pushed the commit with an updated version of `SystemReflectionMetadataVersion`

**Comment by jaredpar** [comment]
> > Not sure if that coordination is still needed, since [sdk is using 8.0](https://github.com/dotnet/sdk/blob/eb47a3cf296746d8624124b5c002db9dc81bbc44/eng/Versions.props#L36) while runtime is at 7.0, and code flow is working fine. @ViktorHofer, thoughts on updating S.R.M to latest alpha?

> cc @jaredpar @rainersigwald

Not quite sure what you're looking for here from us.

**Comment by jaredpar** [comment]
Hard for me to say cause I don't really know what that component is or what it does here.

File: eng/Versions.props
Comment (am11) [inline eng/Versions.props:131]: Same version for `SystemCollectionsImmutableVersion` on line 125.

**Comment by ViktorHofer** [comment]
@am11 there might be dragons, i.e. https://github.com/dotnet/runtime/blob/48ae7e11727d0d2c77e5468cc5aa631816b7eadc/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj#L18-L21

**Comment by am11** [comment]
> I'm not sure if bumping the centrally defined version is the right thing to do.

Yes, we can pin it the other way around (for now ™️); use 9.0-alpha in ILCompiler.Reflection.ReadyToRun.csproj instead.

Maybe we should add something like `SystemReflectionMetadata_CrossRepoSync_Version` for `Microsoft.NET.HostModel.csproj` and `Microsoft.NET.WebAssembly.Webcil.csproj`? Currently the former has hardcoded version and latter is using central.

**Comment by ViktorHofer** [comment]
> Maybe we should add something like SystemReflectionMetadata_CrossRepoSync_Version for Microsoft.NET.HostModel.csproj and Microsoft.NET.WebAssembly.Webcil.csproj? Currently the former has hardcoded version and latter is using central.

Yes I think that would be good. Would you mind opening an issue to track that?

**Comment by rainersigwald** [comment]
The relevant build scenario is:

1. In .NET Framework MSBuild (or VS):
2. Some .NET SDK build tasks use HostModel.
3. HostModel uses core libraries like System.Collections.Immutable and System.Reflection.Metadata.
4. MSBuild also uses those, and provides binding redirects to the VS version.
5. This can cause runtime mismatches when the version resolved in the HostModel context (and bumped through MSBuild binding redirects to generally the last released version) doesn't match the version directly referenced by SDK build tasks (generally that's the current version--9.0.0.0 now).

I think if _everything_ in the SDK references a 9.0 version it'll be fine, the problems come up when, for example, HostModel is old or has a stale reference.

File: eng/Versions.props
Comment (ashaurtaev) [reply eng/Versions.props:131]: Thank you! I changed the version for `SystemCollectionsImmutableVersion`

**Comment by ashaurtaev** [comment]
@am11, how to proceed? Do I need to wait for https://github.com/dotnet/runtime/issues/96809 to complete?

File: eng/Versions.props
Comment (am11) [reply eng/Versions.props:131]: Could you apply this patch:

```diff
diff --git a/eng/Versions.props b/eng/Versions.props
index 9eafadc95df..5a46bfb7ad2 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -122,13 +122,14 @@
     <MicrosoftWin32RegistryVersion>5.0.0</MicrosoftWin32RegistryVersion>
     <StyleCopAnalyzersVersion>1.2.0-beta.507</StyleCopAnalyzersVersion>
     <SystemBuffersVersion>4.5.1</SystemBuffersVersion>
-    <SystemCollectionsImmutableVersion>9.0.0-alpha.1.24064.3</SystemCollectionsImmutableVersion>
+    <SystemCollectionsImmutableVersion>7.0.0</SystemCollectionsImmutableVersion>
     <SystemComponentModelAnnotationsVersion>5.0.0</SystemComponentModelAnnotationsVersion>
     <SystemDataSqlClientVersion>4.8.5</SystemDataSqlClientVersion>
     <SystemDrawingCommonVersion>8.0.0</SystemDrawingCommonVersion>
     <SystemIOFileSystemAccessControlVersion>5.0.0</SystemIOFileSystemAccessControlVersion>
     <SystemMemoryVersion>4.5.5</SystemMemoryVersion>
     <SystemReflectionMetadataVersion>9.0.0-alpha.1.24064.3</SystemReflectionMetadataVersion>
+    <SystemReflectionMetadata_CrossRepoSync_Version>7.0.0</SystemReflectionMetadata_CrossRepoSync_Version>
     <SystemSecurityAccessControlVersion>6.0.0</SystemSecurityAccessControlVersion>
     <SystemSecurityCryptographyCngVersion>5.0.0</SystemSecurityCryptographyCngVersion>
     <SystemSecurityCryptographyOpenSslVersion>5.0.0</SystemSecurityCryptographyOpenSslVersion>
diff --git a/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj b/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj
index cdd3751e3c2..89f6f938536 100644
--- a/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj
+++ b/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj
@@ -20,7 +20,7 @@
 
   <ItemGroup>
     <!-- we need to keep the version of System.Reflection.Metadata in sync with dotnet/msbuild and dotnet/sdk -->
-    <PackageReference Include="System.Reflection.Metadata" Version="6.0.0" />
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadata_CrossRepoSync_Version)" />
     <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
   </ItemGroup>
 
diff --git a/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj b/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj
index d09ae4a569a..677f9dc6150 100644
--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj
@@ -17,7 +17,7 @@
   <ItemGroup>
     <!-- we need to keep the version of System.Reflection.Metadata in sync with dotnet/msbuild and dotnet/sdk -->
     <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" /> 
-    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadata_CrossRepoSync_Version)" />
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
   </ItemGroup>
```
I think this will resolve @ViktorHofer's concern.

**Comment by am11** [comment]
https://github.com/dotnet/runtime/pull/97021#discussion_r1469278711 I think this will address @ViktorHofer's concern, and unblock this PR.

File: eng/Versions.props
Comment (ViktorHofer) [reply eng/Versions.props:131]: Don't you need the latest SCI anymore?

Comment (am11) [reply eng/Versions.props:131]: > Don't you need the latest SCI anymore?

it is only needed by wasm project. Perhaps other consumers of the package would benefit if it’s decoupled as well? (can be done in a separate PR)

Comment (am11) [reply eng/Versions.props:131]: Extracted these changes in a separate PR: https://github.com/dotnet/runtime/pull/97643. @ashaurtaev if that one gets merged first, you can undo the changes here:
`git checkout HEAD~3 -- ':/eng'`
and merge upstream main.

Comment (ashaurtaev) [reply eng/Versions.props:131]: Thank you! I'll wait for your PR to be merged and then rebase my PR

**Comment by ashaurtaev** [comment]
Runtime build for riscv64 completed successfully

**Comment by ashaurtaev** [comment]
@am11 @ViktorHofer Please take a look at this pr since https://github.com/dotnet/runtime/pull/97643 has been merged

**Review by ViktorHofer — COMMENTED** [review]
I'm not the right person to review this change since it touches runtime and the ILCompiler. cc @jkotas @MichalStrehovsky

**Comment by am11** [comment]
> Part of #96438

Should it be `Fixes #96438` or is there something still missing?

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**
